source/reneo.d

```diff
@@ -194,34 +196,43 @@ void sendUTF16OrKeyCombo(wchar unicode_char, bool down) nothrow {
 
     INPUT[] inputs;
 
-    // If Shift modifier is not used: unpress Shift key(s) if already pressed (only for down event).
-    // For Qwertz necessary only for Euro key (AltGr+E) by pressing Shift+7 in Neo layout.
-    bool unpressShift = false;
-    if ((leftShiftDown || rightShiftDown) && !shift && down) {
-        if (leftShiftDown) { appendInput(inputs, VK_LSHIFT, scanLShift, false); }
-        if (rightShiftDown) { appendInput(inputs, VK_RSHIFT, scanRShift, false); }
-        unpressShift = true;
-    }
-
     // For up events, release the main key before the modifiers
     if (!down) {
         appendInput(inputs, vk, Scancode(MapVirtualKey(vk, MAPVK_VK_TO_VSC), false), down);
     }
 
-    // modifiers
-    // pay attention to current capslock state
-    // warning: this is only an approximation. whether capslock affects this key is dependent on the native layout
-    // this means sometimes we need to invert capslock with shift, sometimes not... as of yet this is an open issue.
-    //
-    // Only add Shift in two cases:
-    // (1) Capslock is disabled. If the virtual key combination needs Shift, it will be injected.
-    // (2) Capslock is enabled. Only if the virtual key combination needs no modifiers (Shift/Alt/Ctrl),
-    //     it will be injected. The shift state and the capslock state will cancel each other out.
-    // (2.1) If the key combination requires only Shift, the capslock state works as active Shift key.
-    // (2.2) If higher modifiers Ctrl or Alt are involved, the capslock state has no effect.
-    // So for cases 2.1 and 2.2, no additional Shift is sent.
-    if ((shift && !capslock) || (high == 0 && capslock)) {
-        appendInput(inputs, VKEY.VK_SHIFT, scanLShift, down);
+    // There are three flags that affect the overall result of a current Shift state:
+    // (1) a Shift key is physically pressed
+    // (2) Capslock is active
+    // (3) the key is capslockable
+
+    // The latter can be queried indirectly by reversely generating the Unicode char with a given
+    // keyboard state. If the result of toUnicode() is different from the expected char (and capslock
+    // is active), the key is capslockable.
+    // The other way round, if a key is not capslockable, the Capslock state must not be considered.
+
+    // On capslock disabled, this flag is set to false in both cases (simplifies the XOR later).
+    bool notCapslockable = (buf[0] == unicode_char) && capslock;
+    // Is any Shift key pressed physically?
+    bool shiftDown = leftShiftDown || rightShiftDown;
+    // Calculate the overall Shift state from the flags described, using XOR
+    bool overallShift = shiftDown ^ capslock ^ notCapslockable;
```

Comment:
Sehr gut dokumentiert, die Beweggründe und Logik werden damit klar. Nur die Umsetzung finde ich ohne Wahrheitstabelle schwer lesbar (negierte Variable + Vorverknüpfung von `notCapslockable` mit `capslock` + 3-fach XOR ist ganz schön viel). Mein Vorschlag lautet deshalb wie folgt (angeblich kann man das gleich im Webinterface übernehmen):
```suggestion
    bool nativeCapslockable = buf[0] != unicode_char;
    // Is any Shift key pressed physically?
    bool shiftDown = leftShiftDown || rightShiftDown;
    // Calculate the overall Shift state from the flags described
    bool overallShift = (!nativeCapslockable && shiftDown) || (nativeCapslockable && (shiftDown ^ capslock));
```

Reply:
Ja, die drei Bedingungen (Shift gedrückt, Capslock aktiv, Taste capslock-fähig) wechselwirken miteinander, sind aber nicht gleichwertig. Das macht es schwierig, einen zugleich logischen als auch in Verbindung mit der Definition nachvollziehbaren Ausdruck zu finden.

Deine Variante ist besser in der Hinsicht, dass man keine negierte Variable hat, und dass der Ausdruck für `overallShift` plausibel und nachvollziehbar ist, wenn man eine Minute drüber nachdenkt. Jedoch wird `nativeCapslockable` nicht exakt ermittelt, da man erst bestimmen kann, ob eine Taste capslockable ist, wenn Capslock überhaupt eingeschaltet ist (ansonsten müsste ich ggf. zwei Aufrufe von `toUnicode` durchführen). Logisch genauer wäre es wie bei mir mit dem Nachsatz ` && capslock`.

Mit dieser Prämisse (wenn capslockable, dann auch Capslock aktiv) ergibt sich konsequenterweise für `overallShift`:
- kein Capslock gedrückt → `nativeCapslockable` false → es wird _stets_ der linke Teilausdruck in `overallShift` verwendet
- Capslock gedrückt → man kann den rechten Teilausdruck vereinfachen:
```
bool overallShift = (!nativeCapslockable && shiftDown) || (nativeCapslockable && (shiftDown ^ true));
bool overallShift = (!nativeCapslockable && shiftDown) || (nativeCapslockable && !shiftDown);
bool overallShift = nativeCapslockable ^ shiftDown;
```
Ok, das ist jetzt wieder nicht mehr so gut nachvollziehbar. Aber mit obiger Prämisse eigentlich schon. Fantastisch, wie sich ein hirnzermaterndes Problem in zwei simple Gleichungen aufteilt.

----
In Kurzform: deine Lösung ist deutlich besser zu verstehen, und sie zu kürzen bringt wieder mehr Verwirrung. Ich würde lediglich die erste Zeile wie vorgeschlagen ergänzen: `bool nativeCapslockable = (buf[0] != unicode_char) && capslock;`